### PR TITLE
Switch realm ruleset configuration to use ruleset's `ShortName` as key

### PR DESCRIPTION
**Pull request by peppy**
- [x] Depends on #15739.

Not sure whether we should use `ShortName` or a full relation with `RealmRuleset`. Can revisit in the future if required.

Unfortunate to have to pass in the `RulesetStore`. If only it was static..

### CODE REVIEW

File: osu.Game/OsuGameBase.cs [hunk 1]
```diff
@@ -187,8 +187,9 @@ private void load(ReadableKeyCombinationProvider keyCombinationProvider)
             Resources.AddStore(new DllResourceStore(OsuResources.ResourceAssembly));
 
             dependencies.Cache(contextFactory = new DatabaseContextFactory(Storage));
+            dependencies.Cache(RulesetStore = new RulesetStore(contextFactory, Storage));
```
Comment (bdach) [inline osu.Game/OsuGameBase.cs:190]: After this move, `RulesetStore` now attempts to access DB tables before `runMigrations()` is ran on line 207 in this BDL, causing the rather spectacular CI failure.

**Review by bdach — CHANGES_REQUESTED** [review]

File: osu.Game/OsuGameBase.cs (hunk 1 above)
Comment (peppy) [reply osu.Game/OsuGameBase.cs:190]: I figured something like this would be the case 😅

**Comment by peppy** [comment]
Let me know if the following sits better. I made it as part of my other efforts where I've removed the EF ruleset store completely:

```diff
diff --git a/osu.Game/Database/RealmContextFactory.cs b/osu.Game/Database/RealmContextFactory.cs
index 70ea24b581..9b6919d4c7 100644
--- a/osu.Game/Database/RealmContextFactory.cs
+++ b/osu.Game/Database/RealmContextFactory.cs
@@ -14,7 +14,6 @@
 using osu.Game.Configuration;
 using osu.Game.Input.Bindings;
 using osu.Game.Models;
-using osu.Game.Rulesets;
 using Realms;
 
 #nullable enable
@@ -33,7 +32,7 @@ public class RealmContextFactory : IDisposable, IRealmFactory
         /// </summary>
         public readonly string Filename;
 
-        private readonly RulesetStore? rulesets;
+        private readonly DatabaseContextFactory? efContextFactory;
 
         /// <summary>
         /// Version history:
@@ -77,10 +76,10 @@ public Realm Context
             }
         }
 
-        public RealmContextFactory(Storage storage, string filename, RulesetStore? rulesets = null)
+        public RealmContextFactory(Storage storage, string filename, DatabaseContextFactory? efContextFactory = null)
         {
             this.storage = storage;
-            this.rulesets = rulesets;
+            this.efContextFactory = efContextFactory;
 
             Filename = filename;
 
@@ -252,7 +251,7 @@ private void applyMigrationsForVersion(Migration migration, ulong targetVersion)
                         var newItem = newSettings.ElementAt(i);
 
                         long rulesetId = oldItem.RulesetID;
-                        string? rulesetName = rulesets?.GetRuleset((int)rulesetId)?.ShortName;
+                        string? rulesetName = getRulesetShortNameFromLegacyID(rulesetId);
 
                         if (string.IsNullOrEmpty(rulesetName))
                             migration.NewRealm.Remove(newItem);
@@ -264,6 +263,15 @@ private void applyMigrationsForVersion(Migration migration, ulong targetVersion)
             }
         }
 
+        private string? getRulesetShortNameFromLegacyID(long rulesetId)
+        {
+            if (efContextFactory == null)
+                return null;
+
+            using (var efContext = efContextFactory.Get())
+                return efContext.RulesetInfo.First(r => r.ID == rulesetId)?.ShortName;
+        }
+
         /// <summary>
         /// Flush any active contexts and block any further writes.
         /// </summary>
diff --git a/osu.Game/OsuGameBase.cs b/osu.Game/OsuGameBase.cs
index a2dd417491..1890fd7d24 100644
--- a/osu.Game/OsuGameBase.cs
+++ b/osu.Game/OsuGameBase.cs
@@ -192,7 +192,7 @@ private void load(ReadableKeyCombinationProvider keyCombinationProvider)
 
             dependencies.Cache(RulesetStore = new RulesetStore(contextFactory, Storage));
 
-            dependencies.Cache(realmFactory = new RealmContextFactory(Storage, "client", RulesetStore));
+            dependencies.Cache(realmFactory = new RealmContextFactory(Storage, "client", contextFactory));
 
             dependencies.CacheAs(Storage);
```

**Comment by peppy** [comment]
I've pushed the proposed changes as they will work better going forward (already found other usages in later changes which require this).

**Comment by smoogipoo** [comment]
Crashes at startup:
```
Unhandled exception. System.InvalidOperationException: Sequence contains no elements
   at System.Linq.ThrowHelper.ThrowNoElementsException()
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider.ResultEnumerable`1.GetEnumerator()
   at Microsoft.EntityFrameworkCore.Query.Internal.LinqOperatorProvider._TrackEntities[TOut,TIn](IEnumerable`1 results, QueryContext queryContext, IList`1 entityTrackingInfos, IList`1 entityAccessors)+MoveNext()
   at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Boolean& found)
   at System.Linq.Enumerable.First[TSource](IEnumerable`1 source)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.<>c__DisplayClass15_1`1.<CompileQueryCore>b__0(QueryContext qc)
   at Microsoft.EntityFrameworkCore.Query.Internal.QueryCompiler.Execute[TResult](Expression query)
   at Microsoft.EntityFrameworkCore.Query.Internal.EntityQueryProvider.Execute[TResult](Expression expression)
   at System.Linq.Queryable.First[TSource](IQueryable`1 source, Expression`1 predicate)
   at osu.Game.Database.RealmContextFactory.getRulesetShortNameFromLegacyID(Int64 rulesetId) in /home/smgi/Repos/osu/osu.Game/Database/RealmContextFactory.cs:line 273
   at osu.Game.Database.RealmContextFactory.applyMigrationsForVersion(Migration migration, UInt64 targetVersion) in /home/smgi/Repos/osu/osu.Game/Database/RealmContextFactory.cs:line 260
   at osu.Game.Database.RealmContextFactory.onMigration(Migration migration, UInt64 lastSchemaVersion) in /home/smgi/Repos/osu/osu.Game/Database/RealmContextFactory.cs:line 168
   at Realms.Migration.Execute(Realm oldRealm, Realm newRealm, IntPtr migrationSchema)
--- End of stack trace from previous location ---
   at osu.Framework.Graphics.Containers.CompositeDrawable.loadChild(Drawable child)
   at osu.Framework.Graphics.Containers.CompositeDrawable.load(ShaderManager shaders, Nullable`1 cancellation)
--- End of stack trace from previous location ---
   at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
   at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
   at osu.Framework.Graphics.Containers.CompositeDrawable.loadChild(Drawable child)
   at osu.Framework.Graphics.Containers.CompositeDrawable.load(ShaderManager shaders, Nullable`1 cancellation)
--- End of stack trace from previous location ---
   at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
   at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
   at osu.Framework.Graphics.Containers.CompositeDrawable.loadChild(Drawable child)
   at osu.Framework.Graphics.Containers.CompositeDrawable.load(ShaderManager shaders, Nullable`1 cancellation)
--- End of stack trace from previous location ---
   at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
   at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
   at osu.Framework.Graphics.Containers.CompositeDrawable.loadChild(Drawable child)
   at osu.Framework.Graphics.Containers.CompositeDrawable.load(ShaderManager shaders, Nullable`1 cancellation)
--- End of stack trace from previous location ---
   at osu.Framework.Allocation.BackgroundDependencyLoaderAttribute.<>c__DisplayClass6_0.<CreateActivator>b__4(Object target, IReadOnlyDependencyContainer dc)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyActivator.Activate(Object obj, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Allocation.DependencyContainer.Inject[T](T instance)
   at osu.Framework.Graphics.Drawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Containers.CompositeDrawable.InjectDependencies(IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies)
   at osu.Framework.Graphics.Drawable.Load(IFrameBasedClock clock, IReadOnlyDependencyContainer dependencies, Boolean isDirectAsyncContext)
   at osu.Framework.Platform.GameHost.bootstrapSceneGraph(Game game)
   at osu.Framework.Platform.GameHost.Run(Game game)
   at osu.Desktop.Program.Main(String[] args) in /home/smgi/Repos/osu/osu.Desktop/Program.cs:line 86
```
[client.db.zip](https://github.com/ppy/osu/files/7587155/client.db.zip)

**Comment by smoogipoo** [comment]
It crashes by trying to retrieve a ruleset with ID 4 (doesn't exist in the EF database).

Realm database:
image

**Comment by peppy** [comment]
That's an unfortunate one. I guess we should drop the data where it doesn't exist?

**Comment by smoogipoo** [comment]
That sounds fair.

**Comment by peppy** [comment]
I've pushed a fix for this, can you give it a try again?

**Comment by smoogipoo** [comment]
Looks fine.

My only remaining concern with this is if people rename their rulesets... We really should be assigning GUIDs to rulesets instead ([which can be done via templates](https://github.com/dotnet/dotnet-template-samples/tree/master/14-guid)). But I suppose this can be done later.

**Review by smoogipoo — APPROVED** [review]

**Comment by peppy** [comment]
I think the better solution there is sufficient documentation to ensure a developer doesn't change the short name, since the intention is that is what we will be using when rulesets are submitted online, and is a permanent ID.

**Review by bdach — APPROVED** [review]

**Pull request merged**